dtm0/drlink.c

```diff
@@ -381,30 +385,47 @@ enum dpr_state {
 	DPR_FAILED,
 };
 
-static enum dpr_state dpr_state_infer(struct dtm0_process *proc)
+static enum dpr_state ha_state_infer(const struct m0_confc        *confc,
+				     const struct m0_fid          *fid)
 {
+	struct m0_conf_obj *obj;
+	enum dpr_state      outcome;
+
+	obj = m0_conf_cache_lookup(&confc->cc_cache, fid);
+	M0_ASSERT(obj != NULL);
+
 	/*
-	 * TODO:
-	 * Observe the states of the following enitities:
-	 *	RPC connection
-	 *	RPC session
-	 *	Conf obj
-	 * and then decide whether it is alive, dead or permanently dead.
-	 *
-	 * @verbatim
-	 *	if (conf_obj is ONLINE) {
-	 *		if (conn is ACTIVE && session is in (IDLE, BUSY))
-	 *			return ONLINE;
-	 *		else
-	 *			return TRANSIENT;
-	 *	} else
-	 *		return FAILED;
-	 * @endverbatim
+	 * Drlink should try to connect to those who are able to accept
+	 * incomming connections.
 	 */
-	if (m0_rpc_link_is_connected(&proc->dop_rlink))
-		return DPR_ONLINE;
+	outcome = (M0_IN(obj->co_ha_state,
+			 (M0_NC_ONLINE, M0_NC_DTM_RECOVERING))) ?
+		DPR_ONLINE : DPR_FAILED;
+
+	return M0_RC_INFO(outcome, "fid=" FID_F ", ha_state=%s",
+			  FID_P(fid), m0_ha_state2str(obj->co_ha_state));
+}
+
+
+static enum dpr_state dpr_state_infer(struct m0_dtm0_service *svc,
+				      struct dtm0_process *proc)
+{
+	enum dpr_state  outcome;
+	bool            is_active;
+	struct m0_reqh  *reqh = svc->dos_generic.rs_reqh;
+	struct m0_confc *confc = reqh != NULL ? m0_reqh2confc(reqh) : NULL;
 
-	return DPR_TRANSIENT;
+	outcome = confc != NULL ?
+		ha_state_infer(confc, &proc->dop_rproc_fid) : DPR_TRANSIENT;
+
+	is_active = !proc->dop_rlink.rlk_sess.s_cancelled &&
+		m0_rpc_link_is_connected(&proc->dop_rlink);
+
+	/* Try to reconnect if session is dead but the remote is alive. */
```

Comment:
That's a good question.

It was yet another workaround. There was a situation where RPC link was connected but its session was canceled. If I remember correctly, it happened when the m0client got restarted: when m0client restarted, it went through M0_NC_FAILED which, in turn, canceled the session. In this case, we ended up with ONLINE client, canceled session, and connected (m0_rpc_link_is_connected returns true) rpc link.

At this moment, we are always re-connecting (see `always_reconnect`) to the remote side, so that this whole `dpr_state_infer` thing is redundant at the moment. Besides, I'm not sure if with recent changes in Hare the client goes through FAILED when it restarts. I think we should remove it (https://github.com/ivan-alekhin/cortx-motr/commit/e4eabba54475265477c4b43cd8cf28acad52882a), and address the re-connect issue as a separate improvement for drlink.